File: wallet/src/main/java/com/worldventures/wallet/ui/settings/general/display/impl/DisplayOptionsSettingsPresenterImpl.java
package com.worldventures.wallet.ui.settings.general.display.impl;

import android.support.annotation.NonNull;
import android.support.v4.util.Pair;

import com.worldventures.core.modules.picker.model.PhotoPickerModel;
import com.worldventures.wallet.domain.entity.ImmutableSmartCardUser;
import com.worldventures.wallet.domain.entity.SmartCardUser;
import com.worldventures.wallet.domain.entity.SmartCardUserPhone;
import com.worldventures.wallet.domain.entity.SmartCardUserPhoto;
import com.worldventures.wallet.service.SmartCardInteractor;
import com.worldventures.wallet.service.SmartCardUserDataInteractor;
import com.worldventures.wallet.service.WalletAnalyticsInteractor;
import com.worldventures.wallet.service.WalletSocialInfoProvider;
import com.worldventures.wallet.service.command.SmartCardUserCommand;
import com.worldventures.wallet.service.command.profile.ChangedFields;
import com.worldventures.wallet.service.command.profile.ImmutableChangedFields;
import com.worldventures.wallet.service.command.profile.UpdateSmartCardUserCommand;
import com.worldventures.wallet.service.command.settings.general.display.GetDisplayTypeCommand;
import com.worldventures.wallet.service.command.settings.general.display.SaveDisplayTypeCommand;
import com.worldventures.wallet.ui.common.base.WalletDeviceConnectionDelegate;
import com.worldventures.wallet.ui.common.base.WalletPresenterImpl;
import com.worldventures.wallet.ui.common.navigation.Navigator;
import com.worldventures.wallet.ui.settings.general.display.DisplayOptionsSettingsPresenter;
import com.worldventures.wallet.ui.settings.general.display.DisplayOptionsSettingsScreen;
import com.worldventures.wallet.ui.settings.general.profile.common.ProfileViewModel;
import com.worldventures.wallet.ui.settings.general.profile.common.WalletProfileDelegate;
import com.worldventures.wallet.util.GuaranteedProgressVisibilityTransformer;
import com.worldventures.wallet.util.WalletFilesUtils;

import io.techery.janet.Command;
import io.techery.janet.operationsubscriber.OperationActionSubscriber;
import io.techery.janet.operationsubscriber.view.OperationView;
import io.techery.janet.smartcard.action.settings.SetHomeDisplayTypeAction;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import timber.log.Timber;

public class DisplayOptionsSettingsPresenterImpl extends WalletPresenterImpl<DisplayOptionsSettingsScreen> implements DisplayOptionsSettingsPresenter {

   private final SmartCardInteractor smartCardInteractor;
   private final SmartCardUserDataInteractor smartCardUserDataInteractor;
   private final WalletSocialInfoProvider socialInfoProvider;
   private final WalletProfileDelegate delegate;

   private boolean mustSaveUserProfile;
   private SmartCardUser user;

   public DisplayOptionsSettingsPresenterImpl(Navigator navigator, WalletDeviceConnectionDelegate deviceConnectionDelegate,
         SmartCardInteractor smartCardInteractor, SmartCardUserDataInteractor smartCardUserDataInteractor,
         WalletAnalyticsInteractor analyticsInteractor, WalletSocialInfoProvider socialInfoProvider) {
      super(navigator, deviceConnectionDelegate);
      this.smartCardInteractor = smartCardInteractor;
      this.smartCardUserDataInteractor = smartCardUserDataInteractor;
      this.socialInfoProvider = socialInfoProvider;
      this.delegate = new WalletProfileDelegate(smartCardUserDataInteractor, analyticsInteractor);
   }

   @Override
   public void attachView(DisplayOptionsSettingsScreen view) {
      super.attachView(view);
      initiateData();
      observeHomeDisplay();
      observeUserProfileUploading();
      fetchDisplayType();
   }

   private void initiateData() {
      final ProfileViewModel profileViewModel = getView().getProfileViewModel();
      this.user = (profileViewModel != null) ? delegate.createSmartCardUser(getView().getProfileViewModel()) : null;
      this.mustSaveUserProfile = user != null;
   }

   @SuppressWarnings("ConstantConditions")
   private void observeHomeDisplay() {
      Observable.combineLatest(
            getUserObservable(),
            smartCardInteractor.getDisplayTypePipe().observeSuccess()
                  .map(Command::getResult),
            Pair::new)
            .compose(getView().bindUntilDetach())
            .observeOn(AndroidSchedulers.mainThread())
            .take(1)
            .doOnSubscribe(this::fetchDisplayType)
            .subscribe(pair -> getView().setupViewPager(user, pair.second), t -> Timber.e(t, ""));

      final OperationView<GetDisplayTypeCommand> getDisplayTypeOperationView
            = getView().<GetDisplayTypeCommand>provideGetDisplayTypeOperationView();
      getDisplayTypeOperationView.showProgress(null);
      smartCardInteractor.getDisplayTypePipe().observe()
            .compose(new GuaranteedProgressVisibilityTransformer<>())
            .compose(getView().bindUntilDetach())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(OperationActionSubscriber.forView(getDisplayTypeOperationView)
                  .create()
            );
      smartCardInteractor.saveDisplayTypePipe().observe()
            .compose(new GuaranteedProgressVisibilityTransformer<>())
            .compose(getView().bindUntilDetach())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(OperationActionSubscriber.forView(getView().<SetHomeDisplayTypeAction>provideSaveDisplayTypeOperationView())
                  .onSuccess(command -> {
                     if (getView().getDisplayOptionsSource().isSettings()) {
                        goBack();
                     } else {
                        getNavigator().goBackToProfile();
                     }
                  })
                  .create()
            );
   }

   private void observeUserProfileUploading() {
      delegate.observeProfileUploading(getView());
   }

   @NonNull
   private Observable<SmartCardUser> getUserObservable() {
      return user != null ? Observable.just(user) : smartCardInteractor.smartCardUserPipe()
            .createObservableResult(SmartCardUserCommand.fetch())
            .map(Command::getResult).doOnNext(smartCardUser -> user = smartCardUser);
   }

   @Override
   public void fetchDisplayType() {
      smartCardInteractor.getDisplayTypePipe().send(new GetDisplayTypeCommand(true));
   }

   @Override
   public void savePhoneNumber(ProfileViewModel profile) {
      final SmartCardUserPhone enteredPhone = delegate.createPhone(profile);
      if (enteredPhone == null) {
         return;
      }

      user = ImmutableSmartCardUser.builder()
            .from(user)
            .phoneNumber(enteredPhone)
            .build();

      mustSaveUserProfile = true;
      getView().updateUser(user);
   }

   @Override
   public void handlePickedPhoto(PhotoPickerModel model) {
      getView().cropPhoto(WalletFilesUtils.convertPickedPhotoToUri(model));
   }

   @Override
   public void saveAvatar(String imageUri) {
      user = ImmutableSmartCardUser.builder()
            .from(user)
            .userPhoto(new SmartCardUserPhoto(imageUri))
            .build();

      mustSaveUserProfile = true;
      getView().updateUser(user);
   }

   @Override
   public void choosePhoto() {
      getView().pickPhoto(delegate.provideInitialPhotoUrl(socialInfoProvider.photoThumb()));
   }

   @Override
   public void saveDisplayType(@SetHomeDisplayTypeAction.HomeDisplayType int type) {
      final SaveDisplayTypeCommand saveDisplayType = new SaveDisplayTypeCommand(type, user);
      if (mustSaveUserProfile) {
         updateProfileAndSaveDisplayType(saveDisplayType);
      } else {
         smartCardInteractor.saveDisplayTypePipe().send(saveDisplayType);
      }
   }

   private void updateProfileAndSaveDisplayType(SaveDisplayTypeCommand saveDisplayType) {
      final ChangedFields changedFields = ImmutableChangedFields.builder()
            .firstName(user.firstName())
            .middleName(user.middleName())
            .lastName(user.lastName())
            .phone(user.phoneNumber())
            .photo(user.userPhoto())
            .build();

      smartCardUserDataInteractor.updateSmartCardUserPipe()
            .createObservableResult(new UpdateSmartCardUserCommand(changedFields, true))
            .doOnNext(command -> smartCardInteractor.saveDisplayTypePipe().send(saveDisplayType))
            .subscribe(command -> {
            }, throwable -> Timber.e(throwable, ""));
   }

   @Override
   public void goBack() {
      getNavigator().goBack();
   }
}

File: wallet/src/main/java/com/worldventures/wallet/ui/settings/general/display/impl/DisplayOptionsSettingsPresenterImpl.kt
package com.worldventures.wallet.ui.settings.general.display.impl

import com.worldventures.core.modules.picker.model.PhotoPickerModel
import com.worldventures.wallet.service.SmartCardInteractor
import com.worldventures.wallet.service.WalletSocialInfoProvider
import com.worldventures.wallet.service.command.SmartCardUserCommand
import com.worldventures.wallet.service.command.settings.general.display.GetDisplayTypeCommand
import com.worldventures.wallet.service.command.settings.general.display.SaveDisplayTypeCommand
import com.worldventures.wallet.ui.common.base.WalletDeviceConnectionDelegate
import com.worldventures.wallet.ui.common.base.WalletPresenterImpl
import com.worldventures.wallet.ui.common.navigation.Navigator
import com.worldventures.wallet.ui.settings.general.display.DisplayOptionsSettingsPresenter
import com.worldventures.wallet.ui.settings.general.display.DisplayOptionsSettingsScreen
import com.worldventures.wallet.ui.settings.general.profile.common.WalletProfileDelegate
import com.worldventures.wallet.util.GuaranteedProgressVisibilityTransformer
import com.worldventures.wallet.util.WalletFilesUtils
import io.techery.janet.ActionState
import io.techery.janet.operationsubscriber.OperationActionSubscriber
import io.techery.janet.smartcard.action.settings.SetHomeDisplayTypeAction
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.functions.Action1
import timber.log.Timber

class DisplayOptionsSettingsPresenterImpl(navigator: Navigator,
                                          deviceConnectionDelegate: WalletDeviceConnectionDelegate,
                                          private val delegate: WalletProfileDelegate,
                                          private val smartCardInteractor: SmartCardInteractor,
                                          private val socialInfoProvider: WalletSocialInfoProvider)
   : WalletPresenterImpl<DisplayOptionsSettingsScreen>(navigator, deviceConnectionDelegate), DisplayOptionsSettingsPresenter {

   override fun attachView(view: DisplayOptionsSettingsScreen) {
      super.attachView(view)

      observeHomeDisplay()
      if (!view.isProfileBind) {
         observeUserAndFetchDisplayType(view)
      }
      observeUserProfileUploading()
   }

   private fun observeUserAndFetchDisplayType(view: DisplayOptionsSettingsScreen) {
      Observable.combineLatest(
            smartCardInteractor.smartCardUserPipe()
                  .createObservableResult(SmartCardUserCommand.fetch())
                  .map { delegate.toViewModel(it.result) },
            smartCardInteractor.displayTypePipe
                  .createObservableResult(GetDisplayTypeCommand(true))
                  .map { it.result },
            { first, second -> Pair(first, second) })
            .compose(view.bindUntilDetach())
            .observeOn(AndroidSchedulers.mainThread())
            .take(1)
            .subscribe({ pair -> view.setupDisplayOptions(pair.first, pair.second) }) { t -> Timber.e(t) }
   }

   private fun observeHomeDisplay() {
      smartCardInteractor.displayTypePipe.observe()
            .compose(GuaranteedProgressVisibilityTransformer<ActionState<GetDisplayTypeCommand>>())
            .compose(view.bindUntilDetach())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(OperationActionSubscriber.forView(view.provideGetDisplayTypeOperationView())
                  .create()
            )
      smartCardInteractor.saveDisplayTypePipe().observe()
            .compose(GuaranteedProgressVisibilityTransformer<ActionState<SaveDisplayTypeCommand>>())
            .compose(view.bindUntilDetach())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(OperationActionSubscriber.forView(view.provideSaveDisplayTypeOperationView())
                  .onSuccess {
                     if (!view.isProfileChanged) {
                        proceedWithChangesApply()
                     }
                  }
                  .create()
            )
   }

   private fun observeUserProfileUploading() {
      delegate.observeProfileUploading(view, { proceedWithChangesApply() }, null)
   }

   override fun phoneNumberEntered(phoneCode: String, phoneNumber: String) {
      if (delegate.isPhoneValid(phoneCode, phoneNumber)) {
         view.updatePhone(phoneCode, phoneNumber)
      }
   }

   override fun handlePickedPhoto(model: PhotoPickerModel) {
      view.cropPhoto(WalletFilesUtils.convertPickedPhotoToUri(model))
   }

   override fun avatarSelected(imageUri: String) {
      view.updatePhoto(imageUri)
   }

   override fun choosePhoto() {
      view.pickPhoto(delegate.provideInitialPhotoUrl(socialInfoProvider.photoThumb()))
   }

   override fun saveDisplayType(@SetHomeDisplayTypeAction.HomeDisplayType type: Int) {
      val saveDisplayType = SaveDisplayTypeCommand(type, delegate.createSmartCardUser(view.profile),
            if (view.isProfileChanged) Action1 { delegate.updateUser(it, false) } else null)
      smartCardInteractor.saveDisplayTypePipe().send(saveDisplayType)
   }

   override fun goBack() {
      navigator.goBack()
   }

   override fun fetchDisplayType() {
      smartCardInteractor.displayTypePipe.send(GetDisplayTypeCommand(true))
   }

   private fun proceedWithChangesApply() {
      if (view.displayOptionsSource.isSettings) {
         goBack()
      } else {
         navigator.goBackToProfile()
      }
   }
}
